### ISSUE TITLE
Write a doc and sample on how to use gRPC with nginx

### ISSUE THREAD

**Issue by zinuga** [grpc/grpc.github.io#230]

**Comment by hsaliak** [comment]
@fengli-google would you be able to help out with this?

**Comment by hsaliak** [comment]
@fengli-google have you had a chance to look into this? 
@zinuga is P2 the right priority for this content?

**Comment by zinuga** [comment]
yes, we can drop this from GA milestone.

**Comment by shulegaa** [comment]
Please ...

**Comment by clehene** [comment]
This may be relevant https://trac.nginx.org/nginx/ticket/923

**Comment by zinuga** [comment]
@shulegaa we are in the process of submitting a 3rd party module in nginx for grpc support. we cant control the nginx core code or timeline for http/2 support. for now, if its critical for your needs, we can point you to a custom nginx binary where we have built support for grpc.(usable for testing/prototyping).

**Comment by shulegaa** [comment]
Do tell!  Please.  I'd be very interested in seeing what you've done.  Depending on timing, we may go down a similar path.  Long story.

**Comment by wangming1993** [comment]
I use nginx1.10 which support http2 as grpc server, then go grpc client send a request to server, 
from nginx debug log, I find such message:
```
2016/09/12 14:48:57 [debug] 14677#14677: *1 SSL certificate status callback
2016/09/12 14:48:57 [debug] 14677#14677: *1 SSL NPN advertised

2016/09/12 14:48:57 [debug] 14677#14677: *1 http2 preface verified
2016/09/12 14:48:57 [debug] 14677#14677: *1 process http2 frame type:4 f:0 l:0 sid:0
2016/09/12 14:48:57 [debug] 14677#14677: *1 http2 send SETTINGS frame ack:1
2016/09/12 14:48:57 [debug] 14677#14677: *1 http2 frame complete pos:00007F5DA7C0A5F1 end:00007F5DA7C0A5FE

2016/09/12 14:48:57 [info] 14677#14677: *1 client sent stream with data before settings were acknowledged while processing HTTP/2 connection, client: 127.0.0.1, server: 0.0.0.0:443

```

nginx can receive the connection preface starts with the string `PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n)`, but it send a **RST_STREAM** frame, and refuse to accpet any request.

> Nginx http2 supports required a ack before client send data frame, I know that.

When I use go http2 to send a http request, the nginx can response succeed.   So what's difference and reason for that?  

Anyone helps       

@fengli79 @zinuga @hsaliak @shulegaa @clehene @fengli-google

**Comment by daschwanden** [comment]
@zinuga could you please also point me to your 3rd party nginx moduel for grpc support? thanks

**Comment by fengli79** [comment]
Nginx today doesn't support HTTP2 as upstream, so you cannot use it as a L7 proxy for GRPC. Instead you can use it as a L4 proxy.

For the RST_STREAM issue, it's a know regression in nginx.
https://trac.nginx.org/nginx/ticket/959

**Comment by ynezz** [comment]
@zinuga Hi, I'm also considering grpc and nginx for my next project. Can you please kindly share the source code of that nginx module? I'm not interested in binary blobs :-) Thanks!

@fengli79 BTW that [https://trac.nginx.org/nginx/ticket/959](NGINX RST_STREAM regression) seems to be fixed.

**Comment by hoswey** [comment]
@zinuga could you please also point me to your 3rd party nginx module for grpc support? thanks

**Comment by amitripshtos** [comment]
Much needed feature indeed, I saw some posts on the internet talking about ngx_stream_core_module

**Comment by xiaoyunwu** [comment]
@zinuga same here

**Comment by amitripshtos** [comment]
I went a bit of exploring and understood it better, and created a load balancer that balance automatic using docker + consul:
https://github.com/amitripshtos/grpc-docker-lb

i'm using nghttp2 for this one, by the way

**Comment by ggaaooppeenngg** [comment]
@zinuga  same here, it seems nginx supports HTTP/2, but not sure any example we can check?

**Comment by idevz** [comment]
@zinuga same here, thanks.:)

**Comment by siacomuzzi** [comment]
@zinuga same here (https://github.com/grpc/grpc/issues/11427)

thanks!

**Comment by zinuga** [comment]
based on this https://github.com/grpc/grpc/issues/11427, it is matter of a few weeks now before its upstreamed. credit @PiotrSikora. once done, @LisaFC can help with doc on how to use.

**Comment by siacomuzzi** [comment]
There are some fixes provided by @PiotrSikora in the latest versions of nginx (http://nginx.org/en/CHANGES):

```
Changes with nginx 1.13.2                                        27 Jun 2017

    *) Feature: the "add_trailer" directive.
       Thanks to Piotr Sikora.

    *) Bugfix: in the ngx_http_v2_module.
       Thanks to Piotr Sikora.
```

Do you know if we can use gRPC with nginx now?
Thanks!

**Comment by PiotrSikora** [comment]
@siacomuzzi unfortunately, those are only prerequisites and bugfixes for existing HTTP/2 server-side code. HTTP/2 to backends is still pending review / merge.

**Comment by dyadaval** [comment]
Hello wangming1993, PiotrSikora

Could you please share the working nginx.conf. We are seeing that our nginx h2 proxy(taken from PiotrSikora github) is not forwarding to gRPCServer.  our conf file part looks below:
    server {
    listen 8443 ssl http2;
    listen [::]:8443 ssl http2;
    ssl_certificate /f0/base/nginx/ssl/certs/server.crt;
    ssl_certificate_key /f0/base/nginx/ssl/private/server.key;
    # ...

    location / {
        proxy_pass http://10.220.73.160:50051;
        proxy_set_header X-Real-IP $remote_addr;
        allow all;
    }
10.220.73.160 is the Network element, that has gRPCServer running. 
# nginx -v
nginx version: nginx/1.13.2
#
Thanks for help.

Best Regards,
Div

**Comment by asinha08** [comment]
what is the update on using nginx with grpc?

**Comment by dyadaval** [comment]
No, was not lucky with the patch to get it working. You can use haproxy
with grpc.

**Comment by snowzach** [comment]
So for anyone else here that want's to know the config to make it work I managed to get it going using @PiotrSikora patch. You need to disable buffering in order to make grpc streams work. @PiotrSikora, any idea if they plan to pull this? Pretty happy it works and it would help me out a lot.
```
location /grpcpackagename.grpcservicename {
        proxy_pass https://127.0.0.1:1234;
        proxy_http_version 2.0;
        proxy_pass_trailers on;
        proxy_request_buffering off;
        proxy_buffering off;
}
```

**Comment by pathbox** [comment]
waiting for good news

**Comment by mritunjayblr** [comment]
@snowzach 
I am trying GRPC with nginx, can you please let me know detail of config "location **/grpcpackagename.grpcservicename**"

**Comment by snowzach** [comment]
@mritunjayblr it's basically just that.. GRPC is listening at a url of whatever the GRP package name is followed by the service name... 

So if in your proto definition you have 

package bob

service barker {
rpc blahblahbl;ah
}

your GRPC is technically listening at /bob.barker on your HTTP2 server.

**Comment by mhausenblas** [comment]
If folks want to, happy to contribute the example based on what I did in https://github.com/mhausenblas/kruiser/blob/master/how-does-it-work.md … however, I'm still struggling with serving from non-root URL. So works fine with location `/` but not `/yages.Echo`.

**Comment by derekmma** [comment]
Good news finally come!
https://www.nginx.com/blog/nginx-1-13-10-grpc/

**Comment by dyadaval** [comment]
Hey,

I just signed the petition "Prime Minister of India: We don't want bullet
trains; we want proper bulletproof jackets for our soldiers" and wanted to
see if you could help by adding your name.

Our goal is to reach 187,632 signatures and we need more support. You can
read more and sign the petition here:

http://chng.it/TvbPjd4MYD

Thanks!
Divakar

**Comment by areliszxz** [comment]
si tienen dudas o preguntas chequen https://github.com/areliszxz/nginx-grpc-grpcs es mi primer tuto pero ya pase por todo el show de nginx y grpc-s
